src/nhsuk.base-application/nhsuk.base-application.csproj

```diff
@@ -1,21 +1,91 @@
+    See https://jacobdixon.uk/2020/10/compile-scss-at-build-time-visual-studio
+  -->
+  <ItemGroup>
+    <UpToDateCheckInput Include="wwwroot/src/scss/**/*.scss" Set="css" />
+    <UpToDateCheckBuilt Include="wwwroot/dist/*.css" Set="css" />
+    <UpToDateCheckInput Include="wwwroot/src/js/**/*.js" Set="js" />
+    <UpToDateCheckBuilt Include="wwwroot/dist/*.js" Set="js" />
+    <UpToDateCheckInput Include="wwwroot/src/images/**/*" Set="images" />
+    <UpToDateCheckBuilt Include="wwwroot/dist/images/**/*" Set="images" />
+  </ItemGroup>
+
   <!-- Uncomment to use header and footer Nuget package <ItemGroup>
     <PackageReference Include="NhsUk.header-footer-api-client" Version="2.1.0" />
   </ItemGroup> -->
   <ItemGroup>
     <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
   </ItemGroup>
-  <Target Name="Build frontend" BeforeTargets="BeforeBuild">
+
+  <!--
+    Only run "npm install" if packages.json has been updated since the last time it was run.
+    Will always be run if $(NpmInstallStampFile) does not exist (e.g. the very first build).
+    Deleted output files won't trigger a build. Manually modified output files will not trigger a build.
+    Regeneration of output files can be forced via a rebuild (or clean then build).
+    If using the dotnet command line, the "no-incremental" flag must be set to do a "rebuild".
+    (Based on: https://stackoverflow.com/a/48020233)
+  -->
+  <Target Name="NpmInstall"
+          BeforeTargets="Build frontend"
+          Inputs="package.json"
+          Outputs="$(NpmInstallStampFile)">
     <Exec Command="npm install" />
+    <Touch Files="$(NpmInstallStampFile)" AlwaysCreate="true" />
+  </Target>
+
+  <!--
+    Only build frontend if at least one of the frontend source files was modified since the last frontend build,
+    or if packages.json has been updated since the last frontend build.
+    Will always be run if $(FrontendBuildStampFile) does not exist (e.g. the very first build).
+    Deleted source or output files won't trigger a build. Manually modified output files will not trigger a build.
+    Regeneration of output files can be forced via a rebuild (or clean then build).
+    If using the dotnet command line, the "no-incremental" flag must be set to do a "rebuild".
+  -->
+  <Target Name="Build frontend"
+          BeforeTargets="Build other"
+          Inputs="@(FrontendInputs)"
+          Outputs="$(FrontendBuildStampFile)">
     <Exec Command="npm run build" />
+    <Touch Files="$(FrontendBuildStampFile)" AlwaysCreate="true" />
+  </Target>
+
+  <!-- Moved out of "Build frontend", as assumed this is needed for every build. -->
+  <Target Name="Build other" BeforeTargets="BeforeBuild">
     <ItemGroup>
       <Compile Include="**/*$(DefaultLanguageSourceExtension)" Exclude="$(DefaultItemExcludes);$(DefaultExcludesInProjectFolder);$(BaseIntermediateOutputPath)**;$(BaseOutputPath)**;@(Compile)" />
```

Comment:
I'm not completely sure what this `<Compile ...>` line does, or if it is needed. Kept it to be safe, but it didn't look connected to building frontend so moved it to a new target, named "Build other" for want of a better name.

Reply:
Neither am I to be honest. On an early configuration I had to use it but I think the current method works. I will test VS, CLI, and Docker without it

Reply:
I have found the stack overflow thread which contains this code. I should add that to some comments. 
https://stackoverflow.com/a/44829863/1442943

Reply:
I think I now understand what it is for. The OP on stack overflow is wanting to call an command that generates C# code which needs to be compiled with all the other C# code. The `<Compile ...>` line makes sure the newly generated code is included in the files compiled by the C# compiler.
In this case the command does all the compiling needed and no additional filenames need passing to the C# compiler.
I don't think it is needed here and could be removed.
If keep it, I think best to move it back into the Build frontend target.

Reply:
Removed the "Build other" target.